### PR TITLE
🧹 (Dust to OSMO): Feature implementation

### PR DESCRIPTION
**Pull request by greg-nagy**
## What is the purpose of the change:

This PR is the v0 for the dust to OSMO feature. 

### Outline for the solution

- add a `Dust to OSMO` button to the portfolio page under the asset table 
- when the user clicks `Swap dust to OSMO`
  - loading spinner is added to the button
  - a swap is generated for every asset, and swaps are executed sequentially
  - user is prompted to sign every tx by the wallet
  - after all swap txs resulted in either success or error remove the loading spinner

### Linear Task

<details>
<summary>Brief for the task</summary>

>Currently, a user who regularly trades on the platform may have a number of small balances remaining in their portfolio due to the values of various tokens being incommensurate. These small remaining balances are known as 'dust' and often sit unused since the value is too low in order to justify transacting due to network fees. Your task is to write a function that converts the dust in a wallet (defined as values less than $0.01USD) to OSMO.
>
> Note that there are no designs for this task. You may either write a well tested function for review, or use your best judgement in adding UX to trigger the function.
>
> The task will involve reading and understanding how balances are fetched in the codebase, and understanding signing broadcasting transactions.

</details>

## Brief Changelog

- add a feature flag 
- add the dust to osmo button
- update `useSwap` to be easier to use programatically
- add `getUserDustAsset()` procedure
- on user action execute all the swaps sequentially either with manual signing or 1click trading
- add react query devtools for debugging 
- add a generic query client for non trpc async handling

## Testing and Verifying

This change has been tested locally by rebuilding the website and verified content and links are expected

- set `NEXT_PUBLIC_SHOW_DUST_TO_OSMO=true` in `packages/web/.env.local` 
- go to portfolio 
- make sure you have some dust in the currently connected wallet
- (1click trading off): click `Dust to OSMO` button
- (1click trading on): click `Dust to OSMO` button

## Backlog 

- set up the flag in launch darkly
- drop the environment variable from `.env` and `env.ts` and hardcode a default value in `web/hooks/use-feature-flags.ts`
- Resolve the  `TODO(Greg)` comments 

## To be discussed
- `TODO(Greg)` comments 
- Do we want an overview, confirmation, and error display user interface?
- Should we batch the swaps into a single tx so only 1 signature is necessary?
- Do we want to reduce dust sweep costs? 

### manual sign
CleanShot 2024-11-05 at 09 09 36

### 1click trade
CleanShot 2024-11-05 at 09 15 59

### CODE REVIEW

**Comment by MaxMillington** [comment]
Hi @greg-nagy (apologies for the slow reply). I think for this task, something simpler is in order. A button that triggers swaps for all assets that have values less than $0.01 USD is what we are looking for. The swaps should also be converted to OSMO

**Comment by greg-nagy** [comment]
@MaxMillington I have finished what I had in mind for this PR 

Lloyd emailed me about a bug that could impact the usability of this feature. I’m unsure about the details, but I think it’s worth sharing my progress so you can decide if it adds value. If it does, great; if not, we can close the PR.

File: packages/web/utils/trpc.ts [hunk 1]
```diff
@@ -97,7 +97,7 @@ export const api = createTRPCNext<AppRouter>({
     });
 
     return {
-      queryClient,
+      queryClient: queryClientTRPC,
```
[the remark this answers is not in the document]
Comment (greg-nagy) [reply packages/web/utils/trpc.ts:100]: Valid point, I left a question about this with `TODO(Greg)` in the `_app.tsx` with some thoughts. 

I considered this a good way forward, but we also have other options.

File: packages/trpc/src/assets.ts [hunk 2]
```diff
@@ -408,6 +408,43 @@ export const assetsRouter = createTRPCRouter({
...
+        (
+          asset
+        ): asset is Omit<typeof asset, "amount" | "usdValue"> &
+          Required<Pick<typeof asset, "amount" | "usdValue">> =>
+          asset.amount !== undefined && asset.usdValue !== undefined
+      );
+
+      return assetsWithAmountAndUsdValue.filter((asset) =>
+        asset.usdValue.toDec()?.lte(dustThreshold)
+      );
```
[the remark this answers is not in the document]
Comment (greg-nagy) [reply packages/trpc/src/assets.ts:446]: > The filter assumes usdValue is always valid when defined. Consider adding error handling for potential failures in price fetching.

 There is a filter above this that makes sure this is true. (The filter doesn't check the type assumption that the values are in deed `CoinPretty` or `PricePritty`)

 > Also, consider adding a try-catch block to handle potential price fetching errors:

 The errors are ok to flow through the client.

**Comment by MaxMillington** [comment]
closing for now as some more features are required on the chain side to get this done

**Pull request closed without merging**